droid-results/src/main/java/uk/gov/nationalarchives/droid/profile/datawriter/TemplateBasedDataWriter.java

```diff
@@ -0,0 +1,228 @@
+                        } else {
+                            addNodeColumn(nodeEntries, node, def);
+                        }
+                    }
+                }
+                csvWriter.writeRow(nodeEntries);
+            }
+        }
+        csvWriter.flush();
+    }
+
+    @Override
+    public void writeHeadersForOneRowPerFile(List<? extends ProfileResourceNode> nodes, String[] headers, CsvWriter csvWriter) {
+        super.setCustomisedHeaders(headers);
+        int maxIdCount = getMaxIdentificationCount(nodes);
+        List<String> headersToWrite = getHeadersToWrite(maxIdCount);
+        csvWriter.writeHeaders(headersToWrite);
+        csvWriter.flush();
+    }
+
+    @Override
+    public void writeHeadersForOneRowPerFormat(List<? extends ProfileResourceNode> nodes, String[] headers, CsvWriter csvWriter) {
+        super.setCustomisedHeaders(headers);
+        int maxIdColumns = 1; //for "per format" export, the additional identified formats are written as new row hence no additional id headers
+        List<String> headersToWrite = getHeadersToWrite(maxIdColumns);
+        csvWriter.writeHeaders(headersToWrite);
+        csvWriter.flush();
+    }
+
+    private void addFormatColumnTemplate(List<String> nodeEntries, ProfileResourceNode node, ExportTemplateColumnDef def, int maxIdCount) {
+        List<Format> formats = node.getFormatIdentifications();
+        for (int i = 0; i < maxIdCount; i++) {
+            Format format = i < formats.size() ? formats.get(i) : null;
+            nodeEntries.add(def.getOperatedValue(getFormatValue(def.getOriginalColumnName(), format)));
+        }
+    }
+
+    //CHECKSTYLE:OFF Swith without default
+    private String getFormatValue(String columnName, Format format) {
+        String retVal = CsvWriterConstants.EMPTY_STRING;
+        switch (columnName) {
+            case CsvWriterConstants.HEADER_NAME_PUID:
+                retVal = format == null ? CsvWriterConstants.EMPTY_STRING : format.getPuid();
+                break;
+            case CsvWriterConstants.HEADER_NAME_MIME_TYPE:
+                retVal = format == null ? CsvWriterConstants.EMPTY_STRING : format.getMimeType();
+                break;
+            case CsvWriterConstants.HEADER_NAME_FORMAT_NAME:
+                retVal = format == null ? CsvWriterConstants.EMPTY_STRING : format.getName();
+                break;
+            case CsvWriterConstants.HEADER_NAME_FORMAT_VERSION:
+                retVal = format == null ? CsvWriterConstants.EMPTY_STRING : format.getVersion();
+        }
+        return retVal;
+    }
+    //CHECKSTYLE:ON
+
+    //CHECKSTYLE:OFF - cyclomatic complexity is too high but we have to go through so many column names!!
+    private void addNodeColumn(List<String> nodeEntries, ProfileResourceNode node, ExportTemplateColumnDef def) {
+        NodeMetaData metaData = node.getMetaData();
+
+        String columnValue = "";
+        switch (def.getOriginalColumnName()) {
```

Comment:
Is it worth swapping this out for a map? Something like:

```java
Map<String, String> map = new HashMap<>();
map.put(CsvWriterConstants.HEADER_NAME_ID, nullSafeNumber(node.getId()));
String columnValue = map.get(def.getOriginalColumnName());
```

That would stop it complaning about cyclomatic complexity and would be fewer lines. What do you think?

Reply:
Ooooh HOFs, let me try this out

Reply:
I don't think it would be that simple, the "value" column in a map would need to be an operation which takes "node" as a parameter. If anything I imagine it could be something like 
```
Map<String, SomeOperation> = new HashMap<>()`

interface SomeOperation {
    public String getValue(ProfileResourceNode node); 
}

public class ParentID implements SomeOperation {
    public String getValue(ProfileResourceNode node) {
       nullSafeNumber(node.getId());
    }
}
```

probably too much for what we would gain out of it, what do you think?